jmxtrans-cluster/src/main/java/com/googlecode/jmxtrans/cluster/zookeeper/ZookeeperClusterService.java

```diff
@@ -0,0 +1,228 @@
+import org.apache.curator.retry.ExponentialBackoffRetry;
+import org.apache.curator.utils.ZKPaths;
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.data.Stat;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Created by kulcsart on 5/14/2016.
+ *
+ */
+public class ZookeeperClusterService extends Thread implements ClusterService {
+    private static final Logger log = LoggerFactory.getLogger(ZookeeperClusterService.class);
+    private Configuration configuration;
+    private Injector injector;
+
+    private ZookeeperConfig clConfig;
+    private CuratorFramework clClient;
+
+    PathChildrenCache jvmRootCache;
+    PathChildrenCache workerRootCache;
+
+    Map<String, ZookeeperJvmHandler> jvmHandlers= ImmutableMap.of();
+
+
+    @Inject
+    public ZookeeperClusterService(Injector injector, Configuration configuration) {
+        log.info(this.getClass().getName() + " initiated");
+        checkNotNull(this.injector = injector, "Injector cannot be null");
+        checkNotNull(this.configuration = configuration, "Configuration cannnot be null");
+    }
+
+    private void initilaize() {
+        clConfig = ZookeeperConfigBuilder.buildFromProperties(configuration);
+        try {
+            startClusterConnection();
+            jvmRootCache    = new PathChildrenCache(clClient, clConfig.getConfigPath(), false);
+            workerRootCache = new PathChildrenCache(clClient, clConfig.getHeartBeatPath(),false);
+            jvmRootCache.getListenable().addListener(new JvmConfigPathChangeListener());
+            workerRootCache.getListenable().addListener(new WorkerPathChangeListener());
+        } catch (Exception e) {
+            log.error(Throwables.getStackTraceAsString(e));
+        }
+    }
+
+    private void startClusterConnection() throws Exception {
+        this.clClient = CuratorFrameworkFactory
+                .newClient(clConfig.getConnectionString(),
+                        new ExponentialBackoffRetry(clConfig.getConnectTimeout(), clConfig.getConnectRetry()));
+        clClient.start();
+    }
+
+    @Override
+    public void startService() {
+        initilaize();
+        this.start();
+    }
+
+    @Override
+    public void stopService() {
+        this.interrupt();
+    }
+
+    @Override
+    public void run() {
+        try {
+            startHeartBeating();
+        } catch (Exception e) {
+            log.error(Throwables.getStackTraceAsString(e));
+        }
+        synchronized (this) {
+            this.notifyAll();
+        }
+
+        while (!isInterrupted()) {
+            try {
+                Thread.sleep(1000);
+            } catch (InterruptedException e) {
+                log.error(Throwables.getStackTraceAsString(e));
```

Comment:
If we receive an `InterruptedException`, it is because something wants us to stop working. We should comply. (yes `InterruptedException`s are messy and hard to handle). In this case, you should probably just reset the interrupt state of the trhead and exit the `while` loop.